### ISSUE TITLE
[IGMP] - heap-use-after-free when turning from idle to non-member

### ISSUE THREAD

**Issue by jelledevleeschouwer** [tass-belgium/picotcp#383]
For what it appears there seems to be a ```heap-use-after-free``` in the IGMP-module when a host turns from the _idle-state_ (timer-expired) into _non-member state_ (leaving). This because the same frame is used for the sending of the leaving-report as for the sending of the idle-report. This, while the idle-report frame is already discarded in devloop_out.

Some context:
I have a fully initialized mDNS-module where the module has successfully joined a multicast group after which it becomes idle. Then I want to re-init the mDNS module for some reason where I delete the socket and thus leave the multicast group.

I don't know a lot about IGMP but shouldn't the module not generate a new report for leaving?

This issue relates to #382 

## Debug output
```
Initialised with hostname: hostfoo.local
>>>>>> DONE - CLAIM SESSION: 2
Ethernet sending: 0x60c00000bc80
IGMP: timer expired for FB0000E0 link 0200320A type 1, delay 6248
IGMP: process event on group address FB0000E0
IGMP: state = 1 (0: non-member - 1: delaying member - 2: idle member)
IGMP: event = timer expired | action = send report, set flag
Sending report frame 0x60600000e6c0
IGMP: send membership report on group FB0000E0 to 160000E0
IGMP: new state = idle member
Ethernet sending: 0x60600000e6c0

Try reinitialising mDNS (a second time)
IGMP: process event on group address FB0000E0
IGMP: state = 2 (0: non-member - 1: delaying member - 2: idle member)
IGMP: event = leave group | action = send leave if flag set
Sending report frame 0x60600000e6c0
IGMP: send membership report on group FB0000E0 to 160000E0
IGMP: new state = non-member
```

## Address sanitizer output
```
==19343==ERROR: AddressSanitizer: heap-use-after-free on address 0x60600000e6c6 at pc 0x7ffff6f337eb bp 0x7fffffffda70 sp 0x7fffffff$230
WRITE of size 6 at 0x60600000e6c6 thread T0
    #0 0x7ffff6f337ea (/usr/lib/x86_64-linux-gnu/libasan.so.1+0x2e7ea)
    #1 0x43bfb1 in pico_ethernet_send stack/pico_stack.c:630
    #2 0x4335c3 in devloop_sendto_dev stack/pico_device.c:263
    #3 0x433644 in devloop_out stack/pico_device.c:282
    #4 0x4336cc in devloop stack/pico_device.c:303
    #5 0x4337d0 in pico_devices_loop stack/pico_device.c:348
    #6 0x43d340 in pico_stack_tick stack/pico_stack.c:1091
    #7 0x43f4fe in app_mdns /media/psf/Home/Developer/dev/picotcp/test/examples/mdns.c:66
    #8 0x442880 in main test/picoapp.c:634
    #9 0x7ffff6758b44 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b44)
    #10 0x4013d8 (/media/psf/Home/Developer/dev/picotcp/build/test/picoapp.elf+0x4013d8)

0x60600000e6c6 is located 6 bytes inside of 60-byte region [0x60600000e6c0,0x60600000e6fc)
freed by thread T0 here:
    #0 0x7ffff6f59527 in __interceptor_free (/usr/lib/x86_64-linux-gnu/libasan.so.1+0x54527)
    #1 0x433a68 in pico_frame_discard stack/pico_frame.c:36
    #2 0x433668 in devloop_out stack/pico_device.c:284
    #3 0x4336cc in devloop stack/pico_device.c:303
    #4 0x4337d0 in pico_devices_loop stack/pico_device.c:348
    #5 0x43d340 in pico_stack_tick stack/pico_stack.c:1091
    #6 0x43f4fe in app_mdns /media/psf/Home/Developer/dev/picotcp/test/examples/mdns.c:66
    #7 0x442880 in main test/picoapp.c:634
    #8 0x7ffff6758b44 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b44)

previously allocated by thread T0 here:
    #0 0x7ffff6f59885 in calloc (/usr/lib/x86_64-linux-gnu/libasan.so.1+0x54885)
    #1 0x433bc2 in pico_frame_do_alloc stack/pico_frame.c:88
    #2 0x433cd7 in pico_frame_alloc stack/pico_frame.c:125
    #3 0x41637f in pico_ipv4_alloc modules/pico_ipv4.c:485
    #4 0x413d47 in pico_igmpv3_generate_report modules/pico_igmp.c:706
    #5 0x4140fe in pico_igmp_generate_report modules/pico_igmp.c:792
    #6 0x4141f7 in srsfst modules/pico_igmp.c:839
    #7 0x414a13 in pico_igmp_process_event modules/pico_igmp.c:1090
    #8 0x413ae8 in pico_igmp_state_change modules/pico_igmp.c:633
    #9 0x416a81 in pico_ipv4_mcast_join modules/pico_ipv4.c:768
    #10 0x43a1a5 in mcast_so_addm stack/pico_socket_multicast.c:987
    #11 0x43ae2f in pico_setsockopt_mcast stack/pico_socket_multicast.c:1356
    #12 0x42835a in pico_setsockopt_udp modules/pico_socket_udp.c:230
    #13 0x43787e in pico_socket_setoption stack/pico_socket.c:1785
    #14 0x42332a in pico_mdns_init modules/pico_mdns.c:3563
    #15 0x43f4b6 in app_mdns /media/psf/Home/Developer/dev/picotcp/test/examples/mdns.c:55
    #16 0x442880 in main test/picoapp.c:634
    #17 0x7ffff6758b44 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21b44)
```

**Comment by roel0** [comment]
Correct, a message of the IGMP_LEAVE_TYPE should have been created

fixed in 369bd995dc230c4c415237328fe35cbe4739000e

**Comment by jelledevleeschouwer** [comment]
@roel0 that was fast, thanks! :+1:
